File: core/src/main/java/io/toonformat/toon4s/JToon4s.java
package io.toonformat.toon4s;

import java.io.Reader;
import java.io.Writer;

import scala.util.Either;

/**
 * Java-friendly façade for toon4s.
 * Exposes stdlib-only signatures so Java callers avoid Scala collections in APIs.
 */
public final class JToon4s {
  private JToon4s() {}

  public static Either<io.toonformat.toon4s.error.EncodeError, String> encode(Object value, EncodeOptions options) {
    return Toon.encode(value, options);
  }

  public static Either<io.toonformat.toon4s.error.EncodeError, Void> encodeTo(Object value, Writer out, EncodeOptions options) {
    return Toon.encodeTo(value, out, options).map(u -> null);
  }

  public static Either<io.toonformat.toon4s.error.DecodeError, JsonValue> decode(String toon, DecodeOptions options) {
    return Toon.decode(toon, options);
  }

  public static Either<io.toonformat.toon4s.error.DecodeError, JsonValue> decodeFrom(Reader in, DecodeOptions options) {
    return Toon.decodeFrom(in, options);
  }
}

File: core/src/main/scala/io/toonformat/toon4s/JToon4s.scala
package io.toonformat.toon4s

import java.io.Writer

import io.toonformat.toon4s.error.{DecodeError, EncodeError}

object JToon4s {

  def encode(
      value: Any,
      options: EncodeOptions = EncodeOptions(),
  ): ToonResult[EncodeError, String] =
    Toon.encode(value, options) match {
    case Right(toon) => ToonResult.success(toon)
    case Left(err)   => ToonResult.failure(err)
    }

  def encodeTo(
      value: Any,
      out: Writer,
      options: EncodeOptions = EncodeOptions(),
  ): ToonResult[EncodeError, Void] =
    Toon.encodeTo(value, out, options) match {
    case Right(_)  => ToonResult.success(null)
    case Left(err) => ToonResult.failure(err)
    }

  def decode(
      input: String,
      options: DecodeOptions = DecodeOptions(),
  ): ToonResult[DecodeError, JsonValue] =
    Toon.decode(input, options) match {
    case Right(value) => ToonResult.success(value)
    case Left(err)    => ToonResult.failure(err)
    }

  def decodeFrom(
      in: java.io.Reader,
      options: DecodeOptions = DecodeOptions(),
  ): ToonResult[DecodeError, JsonValue] =
    Toon.decodeFrom(in, options) match {
    case Right(value) => ToonResult.success(value)
    case Left(err)    => ToonResult.failure(err)
    }

}
